File: legacy/v0.0.8/ppf-contact-solver/eigsys/eigsys_2.py
# File: eigsys_2.py
# Code: Claude Code and Codex
# Review: Ryoichi Ando (ryoichi.ando@zozo.com)
# License: Apache v2.0

import numpy as np
import numpy.linalg as LA
from eig_py.eigsolve2x2 import sym_eigsolve_2x2

h = 1e-4
F = np.random.rand(3, 2)

# ==== uncommnent below to try a case two singular values are almost the same =====
# a, b = 0.75, 0.75
# U, _, Vt = LA.svd(F)
# F = U[:, 0:2] @ np.diag([a, b]) @ Vt


def svd3x2(F):
    A = F.T @ F
    lmd, eigvecs = sym_eigsolve_2x2(A)
    sigma = np.sqrt(lmd)
    U = F @ eigvecs
    for i in range(2):
        U[:, i] /= np.linalg.norm(U[:, i])
    return U, sigma, eigvecs.T


def extend(U):
    uv = np.cross(U[:, 0], U[:, 1])
    return np.array([U[:, 0], U[:, 1], uv]).T


def energy_s(a, b):
    return np.log(a**2) + np.log(b**2) + a * b + a**2 * b**2


def energy_F(F):
    _, sigma, _ = svd3x2(F)
    return energy_s(*sigma)


def approx_grad_s(a, b):
    return np.array(
        [
            (energy_s(a + h, b) - energy_s(a - h, b)) / (2 * h),
            (energy_s(a, b + h) - energy_s(a, b - h)) / (2 * h),
        ]
    )


def approx_hess_s(a, b):
    H = np.zeros((2, 2))
    H[:, 0] = (approx_grad_s(a + h, b) - approx_grad_s(a - h, b)) / (2 * h)
    H[:, 1] = (approx_grad_s(a, b + h) - approx_grad_s(a, b - h)) / (2 * h)
    return H


def approx_grad_F(F, dF):
    return (energy_F(F + h * dF) - energy_F(F - h * dF)) / (2 * h)


def approx_hess_F(F, dF):
    H = np.zeros((len(dF), len(dF)))
    for i, dFi in enumerate(dF):
        for j, dFj in enumerate(dF):
            H[i][j] = (
                approx_grad_F(F + h * dFi, dFj) - approx_grad_F(F - h * dFi, dFj)
            ) / (2 * h)
    return H


def gen_dF(i, j):
    dF = np.zeros((3, 2))
    dF[i][j] = 1
    return dF


def mat2vec(A):
    x = []
    for j in range(2):
        for i in range(3):
            x.append(A[i][j])
    return np.array(x)


dF = []
for j in range(2):
    for i in range(3):
        dF.append(gen_dF(i, j))

print("---- numerical gradient ----")
g_F = np.zeros(6)
for i, dFi in enumerate(dF):
    g_F[i] = approx_grad_F(F, dFi)
g_F = np.reshape(g_F, (2, 3)).T
print(g_F)

print("---- numerical hessian ----")
H_F = approx_hess_F(F, dF)
print(H_F)

####### Analytical Eigen Decomposition #######

U, (a, b), Vt = svd3x2(F)
U = extend(U)
print("singular values:", a, b)

H_s, g_s = approx_hess_s(a, b), approx_grad_s(a, b)
S_s, U_s = sym_eigsolve_2x2(H_s)

print("---- analytical gradient ----")
g_F_rebuilt = np.zeros(6)
for i, dFi in enumerate(dF):
    g_F_rebuilt[i] = sum(
        g_s[k] * np.dot(U[:, k], np.dot(dFi, Vt[k, :])) for k in range(2)
    )
g_F_rebuilt = np.reshape(g_F_rebuilt, (2, 3)).T
print(g_F_rebuilt)

Q = [
    np.array([[0, 1], [-1, 0], [0, 0]]) / np.sqrt(2),
    np.array([[0, 1], [1, 0], [0, 0]]) / np.sqrt(2),
    np.array([[0, 0], [0, 0], [1, 0]]),
    np.array([[0, 0], [0, 0], [0, 1]]),
    np.concatenate((np.diag(U_s[:, 0]), np.zeros((1, 2)))),
    np.concatenate((np.diag(U_s[:, 1]), np.zeros((1, 2)))),
]
lmds = [
    (g_s[0] + g_s[1]) / (a + b),
    (g_s[0] - g_s[1]) / (a - b) if abs(a - b) > 1e-2 else H_s[0][0] - H_s[0][1],
    g_s[0] / a,
    g_s[1] / b,
    S_s[0],
    S_s[1],
]

print("--- analytical hessian ---")
Qmat = np.zeros((6, 6))
for i, m in enumerate(Q):
    Qmat[:, i] = mat2vec(U @ m @ Vt)
H_rebuilt = Qmat @ np.diag(lmds) @ Qmat.T
print(H_rebuilt)

###############################################

print("--- error ---")
print("gradient:", LA.norm(g_F - g_F_rebuilt) / LA.norm(g_F))
print("hessian:", LA.norm(H_F - H_rebuilt) / LA.norm(H_F))

File: legacy/v0.0.8/ppf-contact-solver/eigsys/eigsys_2.cpp
// File: eigsys_2.cpp
// Code: Claude Code and Codex
// Review: Ryoichi Ando (ryoichi.ando@zozo.com)
// License: Apache v2.0

#include "eig-hpp/eigsolve2x2.hpp"
#include <Eigen/Dense>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <iostream>
#include <vector>

using namespace Eigen;
static double h = 1e-4;

double energy_s(double a, double b) {
    return std::log(a * a) + std::log(b * b) + a * b + a * a * b * b;
}

double energy_F(const Matrix<double, 3, 2> &F) {
    auto [_U, lambda, _Vt] = run_svd_3x2(F);
    return energy_s(lambda(0), lambda(1));
}

Vector2d approx_grad_s(double a, double b) {
    return Vector2d((energy_s(a + h, b) - energy_s(a - h, b)) / (2 * h),
                    (energy_s(a, b + h) - energy_s(a, b - h)) / (2 * h));
}

Matrix2d approx_hess_s(double a, double b) {
    Matrix2d H;
    H.col(0) = (approx_grad_s(a + h, b) - approx_grad_s(a - h, b)) / (2 * h);
    H.col(1) = (approx_grad_s(a, b + h) - approx_grad_s(a, b - h)) / (2 * h);
    return H;
}

double approx_grad_F(const Matrix<double, 3, 2> &F,
                     const Matrix<double, 3, 2> &dF) {
    return (energy_F(F + h * dF) - energy_F(F - h * dF)) / (2 * h);
}

Matrix<double, 6, 6>
approx_hess_F(const Matrix<double, 3, 2> &F,
              const std::array<Matrix<double, 3, 2>, 6> &dF) {
    int len = dF.size();
    Matrix<double, 6, 6> H = Matrix<double, 6, 6>::Zero(len, len);
    for (int i = 0; i < len; ++i) {
        for (int j = 0; j < len; ++j) {
            H(i, j) = (approx_grad_F(F + h * dF[i], dF[j]) -
                       approx_grad_F(F - h * dF[i], dF[j])) /
                      (2 * h);
        }
    }
    return H;
}

Matrix<double, 3, 2> gen_dF(int i, int j) {
    Matrix<double, 3, 2> dF = Matrix<double, 3, 2>::Zero(3, 2);
    dF(i, j) = 1.0;
    return dF;
}

Vector<double, 6> mat2vec(const Matrix<double, 3, 2> &A) {
    VectorXd x(A.size());
    int index = 0;
    for (int j = 0; j < A.cols(); ++j) {
        for (int i = 0; i < A.rows(); ++i) {
            x(index++) = A(i, j);
        }
    }
    return x;
}

Matrix3d expand_U(const Matrix<double, 3, 2> &U) {
    Vector3d cross = U.col(0).cross(U.col(1));
    Matrix3d result;
    result << U.col(0), U.col(1), cross;
    return result;
}

int main() {
    std::srand(std::time(0));
    std::cout << std::scientific;
    std::cout << std::setprecision(3);

    Matrix<double, 3, 2> F = Matrix<double, 3, 2>::Random(3, 2);
    std::cout << "---- F ----" << std::endl;
    std::cout << F << std::endl;

    std::array<Matrix<double, 3, 2>, 6> dF;
    for (int j = 0; j < 2; ++j) {
        for (int i = 0; i < 3; ++i) {
            dF[i + 3 * j] = gen_dF(i, j);
        }
    }

    Vector<double, 6> g_F = Vector<double, 6>::Zero();
    for (int i = 0; i < 6; ++i) {
        g_F[i] = approx_grad_F(F, dF[i]);
    }
    std::cout << "---- numerical gradient ----" << std::endl;
    std::cout << g_F << std::endl;

    Matrix<double, 6, 6> H_F = approx_hess_F(F, dF);
    std::cout << "---- numerical hessian ----" << std::endl;
    std::cout << H_F << std::endl;

    auto [_U, lambda, Vt] = run_svd_3x2(F);
    Matrix2d H_s = approx_hess_s(lambda(0), lambda(1));
    Vector2d g_s = approx_grad_s(lambda(0), lambda(1));
    auto [S_s, U_s] = sym_eigsolve_2x2(H_s);
    Matrix3d U = expand_U(_U);

    printf("--- analytical gradient ---\n");
    Vector<double, 6> g_F_rebuilt = Vector<double, 6>::Zero();
    for (int i = 0; i < 6; ++i) {
        for (int k = 0; k < 2; ++k) {
            g_F_rebuilt[i] +=
                g_s[k] * U.col(k).dot(dF[i] * Vt.row(k).transpose());
        }
    }
    std::cout << g_F_rebuilt << std::endl;

    // clang-format off
    std::array<Matrix<double, 3, 2>, 6> Q = {
        U * (MatrixXd(3, 2) << 0, 1, -1, 0, 0, 0).finished() * Vt / std::sqrt(2),
        U * (MatrixXd(3, 2) << 0, 1, 1, 0, 0, 0).finished() * Vt / std::sqrt(2),
        U * (MatrixXd(3, 2) << 0, 0, 0, 0, 1, 0).finished() * Vt,
        U * (MatrixXd(3, 2) << 0, 0, 0, 0, 0, 1).finished() * Vt,
        U * (MatrixXd(3, 2) << U_s(0, 0), 0, 0, U_s(1, 0), 0, 0).finished() * Vt,
        U * (MatrixXd(3, 2) << U_s(0, 1), 0, 0, U_s(1, 1), 0, 0).finished() * Vt};
    // clang-format on

    Vector<double, 6> lmds(6);
    lmds << (g_s(0) + g_s(1)) / (lambda(0) + lambda(1)),
        (std::abs(lambda(0) - lambda(1)) > h
             ? (g_s(0) - g_s(1)) / (lambda(0) - lambda(1))
             : H_s(0, 0) - H_s(0, 1)),
        g_s(0) / lambda(0), g_s(1) / lambda(1), S_s(0), S_s(1);

    std::cout << "--- analytical hessian ---" << std::endl;
    Matrix<double, 6, 6> H_rebuilt = Matrix<double, 6, 6>::Zero();
    for (int i = 0; i < 6; ++i) {
        if (lmds[i]) {
            Map<Vector<double, 6>> q(Q[i].data());
            H_rebuilt += lmds[i] * q * q.transpose();
        }
    }
    std::cout << H_rebuilt << std::endl;
    std::cout << "--- error ---" << std::endl;
    std::cout << (g_F - g_F_rebuilt).norm() / g_F.norm() << std::endl;
    std::cout << (H_F - H_rebuilt).norm() / H_F.norm() << std::endl;

    return 0;
}
